mindocr/utils/callbacks.py

```diff
@@ -158,30 +163,42 @@ def __init__(self,
         self.is_main_device = rank_id in [0, None]
         self.loader_eval = loader
         self.network = network
-        self.ema= ema
+        self.ema = ema
         self.logger = print if logger is None else logger.info
         self.val_interval = val_interval
         self.val_start_epoch = val_start_epoch
         self.log_interval = log_interval
         self.batch_size = batch_size
         if self.loader_eval is not None:
-            self.net_evaluator = Evaluator(network, loader, loss_fn, postprocessor, metrics, pred_cast_fp32=pred_cast_fp32, num_columns_to_net=num_columns_to_net, num_columns_of_labels=num_columns_of_labels)
+            self.net_evaluator = Evaluator(network, loader, loss_fn, postprocessor, metrics,
+                                           pred_cast_fp32=pred_cast_fp32, num_columns_to_net=num_columns_to_net,
+                                           num_columns_of_labels=num_columns_of_labels)
             self.main_indicator = main_indicator
             self.best_perf = -1e8
         else:
             self.main_indicator = 'train_loss'
             self.best_perf = 1e8
 
-
         self.ckpt_save_dir = ckpt_save_dir
         if not os.path.exists(ckpt_save_dir):
             os.makedirs(ckpt_save_dir)
 
-        self._losses = list()
         self.last_epoch_end_time = time.time()
         self.epoch_start_time = time.time()
         self.step_start_time = time.time()
 
+        self._losses = []
+        if device_num is not None:
+            reduce_sum = ms.ops.AllReduce()
+
+            @jit
+            def reduce(x):                          # lamda expression is not supported in MindSpore
+                return reduce_sum(x) / device_num   # average loss across all cards
+
```

Comment:
it is better to put reduce outside the function

Reply:
Talked to Jun about it, he said that: 1. callbacks are always executed in native mode, 2. `ops.AllReduce()` may take a noticeable amount of time in native mode due to some overhead computations, so it is generally recommended to wrap it with `jit`.

Although, I agree that reducing single number tensors may be very quick and `jit` could be an overkill here. Maybe we can benchmark later and see if it is really necessary.